GravatarApp/ProfileEditor/EditProfileViewModel.swift

```diff
@@ -1,18 +1,68 @@
+import Combine
 import Gravatar
 import SwiftUI
 
 @MainActor
 class EditProfileViewModel: ObservableObject {
-    private let authToken: String
-    let profile: Profile
+    private let profileService: ProfileService
+    private let userSession: UserSession
+    private var cancellables = Set<AnyCancellable>()
+
+    @Published var isSaving: Bool = false
+    @Published var fields: ProfileFieldsModel
+
+    var isSavinDisabled: Bool {
+        !hasUnsavedChanges || isSaving
+    }
+
+    var hasUnsavedChanges: Bool {
+        guard let profile = userSession.profile else {
+            return false
+        }
+        return !fields.isEqual(to: profile)
+    }
 
     init(
-        profile: Profile,
-        authToken: String
+        userSession: UserSession = .shared,
+        urlSession: URLSessionProtocol? = nil
     ) {
-        self.profile = profile
-        self.authToken = authToken
+        self.userSession = userSession
+        self.profileService = .init(urlSession: urlSession)
+        if let profile = userSession.profile {
+            self.fields = .init(profile: profile)
+        } else {
+            self.fields = .init()
+        }
+        userSession.$profile.sink { [weak self] profile in
+            guard let profile else { return }
+            self?.fields = .init(profile: profile)
+        }
+        .store(in: &cancellables)
     }
 
     // TODO: Implement
+
+    func save() async {
+        guard let authToken = userSession.accessToken else { return }
+        defer {
+            isSaving = false
+        }
+        do {
+            isSaving = true
+            let request = fields.updateRequest()
+            let profile = try await profileService.updateProfile(with: request, token: authToken)
+            print("success!")
+            Task { @MainActor in
+                self.userSession.updateProfile(profile)
+            }
+            // TODO: Show success toast
+        } catch APIError.responseError(let .invalidHTTPStatusCode(response, _))
+            where response.statusCode == HTTPStatus.unauthorized.rawValue
+        {
+            NotificationCenter.default.post(name: .sessionExpired, object: nil)
```

Comment:
If, in the future, it turns out that we always handle session expiration the same way we can move posting of this notification in a deeper and a single place at the networking level.

Reply:
Yes, I agree it would be good to handle this in a centralized way, to avoid the need to check for this error on every single request. We can explore this later on